README.rst

```diff
@@ -1962,6 +1968,347 @@ However, if specifying :code:`Table` objects manually, please make sure to inclu
+
+The compiler's macro system empowers users *reshaping* their data's structure to fit
+their mental model, minimizing query complexity and repetitiveness without requiring changes
+to the shape of the data in the underlying data systems. The compiler achieves this by allowing
+users to define **macros** -- type-safe rules for programmatic query rewriting
+that transform user-provided queries on the *desired* data model into
+queries on the *actual* data model in the underlying data systems.
+
+When macros are defined, the compiler loads them into a `macro registry <Macro registry>`_ -- a
+data structure that tracks all currently available macros, the resulting GraphQL schema
+(accounting for macros), and any additional metadata needed by the compiler.
+The compiler then leverages this registry to expand queries that rely on macros,
+rewriting them into equivalent queries that do not contain any macros and therefore
+reflect the actual underlying data model.
+
+This makes macros somewhat similar to SQL's idea of non-materialized views,
+though there are some key differences:
+
+- SQL views require database access and special permissions; databases are
+  completely oblivious to the use of macros since by the time the database gets the query,
+  all macro uses have been already expanded.
+
+- Macros can be stored and expanded client-side, so different users that query the same system may
+  define their own personal macros which are not shared with other users or the server that executes
+  the users' GraphQL queries. This is generally not achievable with SQL.
+
+- Since macro expansion does not interact in any way with the underlying data system, it works
+  seamlessly with all databases and even on schemas stitched together from multiple databases.
+  In contrast, not all databases support SQL-like :code:`VIEW` functionality.
+
+Currently, the compiler supports one type of macro: `macro edges <Macro edges>`_, which allow
+the creation of "virtual" edges computed from existing ones.
+More types of macros are coming in the future.
+
+Macro registry
+~~~~~~~~~~~~~~
+
+The macro registry is where the definitions of all currently defined macros are stored,
+together with the resulting GraphQL schema they form, as well as any associated metadata
+that the compiler's macro system may need in order to expand any macros encountered in a query.
+
+To create a macro registry object for a given GraphQL schema, use the :code:`create_macro_registry`
+function:
+
+.. code:: python
+
+    from graphql_compiler.macros import create_macro_registry
+
+    macro_registry = create_macro_registry(your_graphql_schema_object)
+
+To retrieve the GraphQL schema object with all its macro-based additions, use
+the :code:`get_schema_with_macros` function:
+
+.. code:: python
+
+    from graphql_compiler.macros import get_schema_with_macros
+
+    graphql_schema = get_schema_with_macros(macro_registry)
+
+Schema for defining macros
+~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
Since I am not sure what a user would do with the macro-definition schema
.

Reply:
You actually can't define macros at all without it -- they won't validate since the macro directives aren't in the usual schema. I consider that a pretty critical detail :) Perhaps I should make that clear in this section.